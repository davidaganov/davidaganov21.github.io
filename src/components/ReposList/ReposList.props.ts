import { DetailedHTMLProps, HTMLAttributes } from "react"
import { ReposProps } from "../../interfaces"

export interface ReposListProps
  extends DetailedHTMLProps<HTMLAttributes<HTMLLIElement>, HTMLLIElement> {
  repos?: ReposProps[]
}
