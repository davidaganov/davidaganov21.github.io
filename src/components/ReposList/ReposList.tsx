import { useState, useEffect } from "react"
import { useTranslation } from "react-i18next"
import { ReposListProps } from "./ReposList.props"
import styles from "./ReposList.module.sass"
import cn from "classnames"

import { CardRepo } from ".."
import { ReposProps } from "../../interfaces"

export const ReposList = ({ repos }: ReposListProps) => {
  const [selectTag, setSelectTag] = useState<string>("")
  const [sortRepos, setSortRepos] = useState<ReposProps[]>([])

  const { t } = useTranslation()

  useEffect(() => {
    if (repos) {
      setSortRepos(repos.filter((item) => item.topics.some((i) => [selectTag].includes(i))))
    }
  }, [selectTag])

  const renderTags = () => {
    let topics: string[] = []

    if (repos) {
      topics = [
        ...new Set(repos.reduce((topics, repo) => [...topics, ...repo.topics], [] as string[]))
      ]
    }

    const tags = topics.map((tag) => (
      <button
        key={tag}
        id={`tag-${tag}`}
        type="button"
        className={cn(styles.tag, {
          [styles.tagActive]: tag === selectTag
        })}
        aria-selected={tag === selectTag}
        aria-labelledby={`sortLabel tag-${tag}`}
        lang="en"
        onClick={() => setSelectTag(tag === selectTag ? "" : tag)}
        role="option"
      >
        {tag}
      </button>
    ))

    if (tags.length !== 0) {
      return (
        <div
          className={styles.tags}
          role="listbox"
        >
          <span
            className="visually-hidden"
            id="sortLabel"
          >
            {t("projects.sort")}
          </span>
          {tags}
        </div>
      )
    }
  }

  const renderItems = () => {
    const arr = sortRepos.length === 0 && repos ? repos : sortRepos
    if (repos) {
      const items = arr.slice(0, 6).map((item) => {
        return (
          <CardRepo
            key={item.id}
            card={item}
          />
        )
      })

      return <ul className={styles.repos}>{items}</ul>
    }
  }

  const renderError = () => {
    return (
      <>
        <p className={styles.empty}>{t("projects.errorTitle")}</p>
        <p className={styles.emptySubtitle}>{t("projects.errorSubtitle")}</p>
      </>
    )
  }

  const tags = renderTags()
  const projects = renderItems()
  const error = projects === undefined ? renderError() : null

  return (
    <div className={styles.container}>
      {tags}
      {projects}
      {error}
    </div>
  )
}
