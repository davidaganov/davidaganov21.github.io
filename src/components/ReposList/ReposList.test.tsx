import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { ReposList } from "./ReposList"

const data = [
  {
    id: "1",
    name: "First",
    description: "Lorem ipsum",
    homepage: "demo",
    html_url: "repo",
    topics: ["vue", "typescript"]
  },
  {
    id: "2",
    name: "Second",
    description: "Dolor sit",
    homepage: "",
    html_url: "repo",
    topics: ["react"]
  }
]

const topics_qty = data.map((i) => i.topics.length).reduce((a, b) => a + b)

describe("ReposList component", () => {
  it("ReposList renders", () => {
    render(<ReposList repos={data} />)

    expect(screen.getAllByRole("option")).toHaveLength(topics_qty)
    expect(screen.getAllByRole("heading")).toHaveLength(data.length)
  })

  it("ReposList renders without data", () => {
    render(<ReposList />)

    expect(screen.queryByRole("option")).toBeNull()
    expect(screen.queryByRole("heading")).toBeNull()
    expect(screen.getByText(/projects not found/i)).toBeInTheDocument()
  })

  it("Sorting repositories is working", () => {
    render(<ReposList repos={data} />)

    expect(screen.getByText(/first/i)).toBeInTheDocument()
    expect(screen.getByText(/second/i)).toBeInTheDocument()

    userEvent.click(screen.getByRole("option", { name: /sort projects by tag react/i }))

    expect(screen.queryByText(/first/i)).toBeNull()
    expect(screen.queryByText(/second/i)).toBeInTheDocument()
  })

  it("ReposList snapshot", () => {
    const repos = render(<ReposList repos={data} />)

    expect(repos).toMatchSnapshot()
  })

  it("ReposList snapshot without data", () => {
    const repos = render(<ReposList />)

    expect(repos).toMatchSnapshot()
  })
})
