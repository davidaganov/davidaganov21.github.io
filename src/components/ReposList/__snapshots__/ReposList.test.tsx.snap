// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repos component Repos snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="tags"
          role="listbox"
        >
          <span
            class="visually-hidden"
            id="sortLabel"
          >
            Sort projects by tag
          </span>
          <button
            aria-labelledby="sortLabel tag-vue"
            aria-selected="false"
            class="tag"
            id="tag-vue"
            role="option"
            type="button"
          >
            vue
          </button>
          <button
            aria-labelledby="sortLabel tag-typescript"
            aria-selected="false"
            class="tag"
            id="tag-typescript"
            role="option"
            type="button"
          >
            typescript
          </button>
          <button
            aria-labelledby="sortLabel tag-react"
            aria-selected="false"
            class="tag"
            id="tag-react"
            role="option"
            type="button"
          >
            react
          </button>
        </div>
        <ul
          class="projects"
        >
          <li
            class="project"
          >
            <h3
              class="title"
            >
              First
            </h3>
            <p
              class="description"
            >
              Lorem ipsum
            </p>
            <div
              class="bottom"
            >
              <a
                aria-label="Go to repository First"
                class="link inline-link"
                href="repo"
                rel="noreferrer"
                target="_blank"
              >
                Go to repository
              </a>
              <a
                aria-label="Live demo First"
                class="link link--live inline-link"
                href="demo"
                rel="noreferrer"
                target="_blank"
              >
                Live demo
              </a>
              <ul
                class="tags"
              >
                <li>
                  vue
                </li>
                <li>
                  typescript
                </li>
              </ul>
            </div>
          </li>
          <li
            class="project"
          >
            <h3
              class="title"
            >
              Second
            </h3>
            <p
              class="description"
            >
              Dolor sit
            </p>
            <div
              class="bottom"
            >
              <a
                aria-label="Go to repository Second"
                class="link inline-link"
                href="repo"
                rel="noreferrer"
                target="_blank"
              >
                Go to repository
              </a>
              <ul
                class="tags"
              >
                <li>
                  react
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="tags"
        role="listbox"
      >
        <span
          class="visually-hidden"
          id="sortLabel"
        >
          Sort projects by tag
        </span>
        <button
          aria-labelledby="sortLabel tag-vue"
          aria-selected="false"
          class="tag"
          id="tag-vue"
          role="option"
          type="button"
        >
          vue
        </button>
        <button
          aria-labelledby="sortLabel tag-typescript"
          aria-selected="false"
          class="tag"
          id="tag-typescript"
          role="option"
          type="button"
        >
          typescript
        </button>
        <button
          aria-labelledby="sortLabel tag-react"
          aria-selected="false"
          class="tag"
          id="tag-react"
          role="option"
          type="button"
        >
          react
        </button>
      </div>
      <ul
        class="projects"
      >
        <li
          class="project"
        >
          <h3
            class="title"
          >
            First
          </h3>
          <p
            class="description"
          >
            Lorem ipsum
          </p>
          <div
            class="bottom"
          >
            <a
              aria-label="Go to repository First"
              class="link inline-link"
              href="repo"
              rel="noreferrer"
              target="_blank"
            >
              Go to repository
            </a>
            <a
              aria-label="Live demo First"
              class="link link--live inline-link"
              href="demo"
              rel="noreferrer"
              target="_blank"
            >
              Live demo
            </a>
            <ul
              class="tags"
            >
              <li>
                vue
              </li>
              <li>
                typescript
              </li>
            </ul>
          </div>
        </li>
        <li
          class="project"
        >
          <h3
            class="title"
          >
            Second
          </h3>
          <p
            class="description"
          >
            Dolor sit
          </p>
          <div
            class="bottom"
          >
            <a
              aria-label="Go to repository Second"
              class="link inline-link"
              href="repo"
              rel="noreferrer"
              target="_blank"
            >
              Go to repository
            </a>
            <ul
              class="tags"
            >
              <li>
                react
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Repos component Repos snapshot without data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <p
          class="empty"
        >
          Projects not found
        </p>
        <p
          class="emptySubtitle"
        >
          Perhaps you should check for github?
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <p
        class="empty"
      >
        Projects not found
      </p>
      <p
        class="emptySubtitle"
      >
        Perhaps you should check for github?
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
